rslib/src/typeanswer.rs

```diff
@@ -5,113 +5,160 @@ use std::borrow::Cow;
+                    expected_tokens.push(DiffToken::missing(expected_str));
                 }
                 "replace" => {
-                    provided.push(DiffToken::bad(self.slice_provided(&opcode)));
-                    expected.push(DiffToken::missing(self.slice_expected(&opcode)));
+                    provided_tokens.push(DiffToken::bad(self.slice_provided(&opcode)));
+                    expected_tokens.push(DiffToken::missing(self.slice_expected(&opcode)));
                 }
                 _ => unreachable!(),
             }
         }
-        DiffOutput { provided, expected }
+        DiffTokens {
+            provided_tokens,
+            expected_tokens,
+        }
     }
 
-    fn to_html(&self) -> String {
-        let output = self.to_tokens();
-        let provided = render_tokens(&output.provided);
-        let expected = render_tokens(&output.expected);
-        format!(
-            "<code id=typeans>{}</code>",
-            if self.provided.is_empty() {
-                htmlescape::encode_minimal(&self.expected.iter().collect::<String>())
-            } else if self.provided == self.expected {
-                provided
-            } else {
-                format!("{provided}<br><span id=typearrow>&darr;</span><br>{expected}")
-            }
-        )
+    // Utility Functions
+    fn prepare_expected(expected: &str) -> String {
+        let no_av_tags = strip_av_tags(expected);
+        let no_linebreaks = LINEBREAKS.replace_all(&no_av_tags, " ");
+        strip_html(&no_linebreaks).trim().to_string()
+    }
+
+    fn get_slice(&self, chars: &[char], start: usize, end: usize) -> String {
+        chars[start..end].iter().collect()
+    }
+
+    fn slice_expected(&self, opcode: &Opcode) -> String {
+        self.get_slice(&self.expected, opcode.second_start, opcode.second_end)
+    }
+
+    fn slice_provided(&self, opcode: &Opcode) -> String {
+        self.get_slice(&self.provided, opcode.first_start, opcode.first_end)
     }
-}
 
-fn prepare_expected(expected: &str) -> String {
-    let without_av = strip_av_tags(expected);
-    let without_newlines = LINEBREAKS.replace_all(&without_av, " ");
-    let without_html = strip_html(&without_newlines);
-    let without_outer_whitespace = without_html.trim();
-    normalize_to_nfc(without_outer_whitespace).into()
+    // Render Functions
+    fn render_tokens(tokens: &[DiffToken]) -> String {
+        tokens.iter().fold(String::new(), |mut acc, token| {
+            let isolated_text = Self::isolate_leading_mark(&token.text);
+            let encoded_text = htmlescape::encode_minimal(&isolated_text);
+            let class = token.to_class();
+            acc.push_str(&format!("<span class={class}>{encoded_text}</span>"));
+            acc
+        })
+    }
+
+    /// Prefixes a leading mark character with a non-breaking space to prevent
+    /// it from joining the previous token.
+    fn isolate_leading_mark(text: &str) -> Cow<str> {
```

Comment:
This method only depends on its inputs and not on self, so is better implemented as a free function. Same for the other methods that don't use self.

Reply:
That might be my familiarity with C++ and its tenants of *"beware namespace pollution"* and encapsulation. But this is Rust, are free functions a convention? The `fn new` "constructors" are also free functions?

Generally I find code easier to read if it's structured following the execution flow as close as possible—the better example for this question here being `fn get_slice`. I also found this way cleaner to rewrite `impl Diff` as shared `trait` in PR #3422, for instance `fn render_expected_tokens` there needs `&self` (in the NonCombining variant) but `fn render_tokens` doesn't.

Reply:
On second thought, as far as this PR is concerned, turning them back into free functions requires barely more than shifting a line.

But: How shall I do it for the eponymous associated functions in the follow-up PR's `trait Diff`'s `impl`? Leave them as-is—it's a shared `trait`, after all? Or append `_noncombining` to the new variants? Since they're not used in the in the trait's default implementations, I can do either here.

Reply:
Done, with the follow-up PR adjusted as these reverts to free functions require.

(Plus two more commits there as suggestions.)